File: Day14.lua
function Split(str)
    local a, b = string.find(str, " -> ")
    return string.sub(str, 1, a-3), string.sub(str, b+1)
end

local filename = "inputs/Day14.in"
local file = io.open(filename, "r")
io.input(file)

local polymer = nil
local rules = {}
for l in io.lines(filename) do
    if (string.len(l) > 0) then
        if (polymer == nil) then
            polymer = l
        else
            table.insert(rules, l)
            local key, value = Split(l)
            rules[key] = value
        end
    end
end

for k=1,40 do
    local new_polymer = ""
    for i=1,string.len(polymer)-1 do
        local first = string.sub(polymer, i, i)
        local second = string.sub(polymer, i+1, i+1)
        local pair = first .. second
        if (rules[pair] ~= nil) then
            new_polymer = new_polymer .. first .. rules[pair]
        else
            new_polymer = new_polymer .. first
        end
    end
    polymer = new_polymer .. string.sub(polymer, #polymer)
end

local counts = {}
for elem in polymer:gmatch(".") do
    if (counts[elem] ~= nil) then
        counts[elem] = counts[elem] + 1
    else
        counts[elem] = 0
    end
end

local mx = 0
local mn = 1000000000000
for _, count in pairs(counts) do
    if mx < count then
        mx = count
    elseif mn > count then
        mn = count
    end
end

print(mx-mn)

File: Day14.ts
import * as fs from 'fs'
import { performance } from 'perf_hooks'

function getCountDiff(_pairs: Map<string, number>, _counts: Map<string, number>, rules: Map<string, string>, limit: number) {
    let pairs: Map<string, number> = new Map<string, number>(_pairs)
    let counts: Map<string, number> = new Map<string, number>(_counts)
    for (let i = 0; i < limit; ++i) {
        let temp_pairs: Map<string, number> = new Map<string, number>()
        let temp_counts: Map<string, number> = new Map<string, number>()
        for (let [pair, count] of pairs) {
            if (rules.has(pair)) {
                let p1 = pair[0] + rules.get(pair)!
                let p2 = rules.get(pair) + pair[1]!
    
                let c1 = count
                if (temp_pairs.has(p1))
                    c1 += temp_pairs.get(p1)!
                temp_pairs.set(p1, c1)
                
                let c2 = count
                if (temp_pairs.has(p2))
                    c2 += temp_pairs.get(p2)!
                temp_pairs.set(p2, c2)
        
                let temp_count = count
                if (temp_counts.has(rules.get(pair)!))
                    temp_count += temp_counts.get(rules.get(pair)!)!
                temp_counts.set(rules.get(pair)!, temp_count)
            } else {
                temp_pairs.set(pair, count)
            }
        }
        for (let [elem, count] of temp_counts) {
            if (counts.has(elem))
                counts.set(elem, counts.get(elem)! + count)
            else
                counts.set(elem, count)
        }
    
        pairs = temp_pairs
    }

    let max = 0
    for (let count of counts.values()) {
        max = count > max ? count : max;
    }

    let min = max
    for (let count of counts.values()) {
        min = count < min ? count : min;
    }

    return max - min;
}

let startTime = performance.now();

let file: string[] = fs.readFileSync('inputs/Day14.in', 'utf-8').split('\n')
let polymer: string = file[0]

let rules: Map<string, string> = new Map<string, string>()
for (let line of file.splice(1)) {
    if (line.length > 0) {
        let kv: string[] = line.split(" -> ")
        rules.set(kv[0], kv[1])
    }
}

let counts: Map<string, number> = new Map<string, number>()
for (let c of polymer) {
    if (!counts.has(c))
        counts.set(c, 0)
    counts.set(c, counts.get(c)! + 1)
}

let pairs: Map<string, number> = new Map<string, number>()
for (let i = 0; i < polymer.length - 1; ++i) {
    let pair = polymer.substr(i, 2)
    if (!pairs.has(pair))
        pairs.set(pair, 0)
    pairs.set(pair, pairs.get(pair)! + 1)
}

let part1: number = getCountDiff(pairs, counts, rules, 10)
let part2: number = getCountDiff(pairs, counts, rules, 40)

let endTime = performance.now();

console.log(part1)
console.log(part2)
console.log(`Time: ${(endTime - startTime).toFixed(3)}ms`)
